subsys/nrf_rpc/nrf_rpc_ipc.c

```diff
@@ -0,0 +1,255 @@
+		}
+		break;
+	}
+
+	return 0;
+}
+
+static void ept_bound(void *priv)
+{
+	const struct nrf_rpc_tr *transport = priv;
+	struct nrf_rpc_ipc *ipc_config = transport->ctx;
+
+	LOG_DBG("nRF RPC endpoint %s connected", ipc_config->endpoint.ept_cfg.name);
+
+	k_event_set(&ipc_config->endpoint.ept_bond, 0x01);
+}
+
+static void ept_received(const void *data, size_t len, void *priv)
+{
+	int err;
+	const struct nrf_rpc_tr *transport = priv;
+	struct nrf_rpc_ipc *ipc_config = transport->ctx;
+
+	__ASSERT_NO_MSG(data != NULL);
+	__ASSERT_NO_MSG(ipc_config->receive_cb != NULL);
+
+	DUMP_LIMITED_DBG(data, len, "Received");
+
+	err = ipc_service_hold_rx_buffer(&ipc_config->endpoint.ept, (void *)data);
+	if (err) {
+		LOG_ERR("ipc_service_hold_rx_buffer failed with err: %d", err);
+		__ASSERT_NO_MSG(false);
+		return;
+	}
+
+	ipc_config->receive_cb(transport, data, len, ipc_config->context);
+}
+
+static void ept_error(const char *message, void *priv)
+{
+	LOG_ERR("Endpoint error: \"%s\"", log_strdup(message));
+	__ASSERT_NO_MSG(false);
+}
+
+static int init(const struct nrf_rpc_tr *transport, nrf_rpc_tr_receive_handler_t receive_cb,
+		void *context)
+{
+	int err;
+	struct nrf_rpc_ipc *ipc_config = transport->ctx;
+	struct nrf_rpc_ipc_endpoint *endpoint = &ipc_config->endpoint;
+	struct ipc_ept_cfg *cfg = &ipc_config->endpoint.ept_cfg;
+
+	if (ipc_config->used) {
+		return 0;
+	}
+
+	if (receive_cb == NULL) {
+		LOG_ERR("No transport receive callback");
+		return -EINVAL;
+	}
+
+	err = ipc_service_open_instance(ipc_config->ipc);
+	if (err && err != -EALREADY) {
+		LOG_ERR("IPC service instance initialization failed: %d\n", err);
+		return translate_error(err);
+	}
+
+	ipc_config->receive_cb = receive_cb;
+	ipc_config->context = context;
+
+	cfg->cb.bound = ept_bound;
+	cfg->cb.received = ept_received;
+	cfg->cb.error = ept_error;
+	cfg->priv = (void *)transport;
+
+	k_event_init(&endpoint->ept_bond);
+
+	err = ipc_service_register_endpoint(ipc_config->ipc, &endpoint->ept, cfg);
+	if (err) {
+		LOG_ERR("Registering endpoint failed with %d", err);
+		return translate_error(err);
+	}
+
+	ipc_config->used = true;
+
+	if (!k_event_wait(&endpoint->ept_bond, 0x01, false,
+			  EPT_BIND_TIMEOUT)) {
+		ipc_config->used = false;
+
+		LOG_ERR("IPC endpoint bond timeout");
+		return -EPIPE;
+	}
+
+	return 0;
+}
+
+int send(const struct nrf_rpc_tr *transport, const uint8_t *data, size_t length)
+{
+	int err;
+	int err2;
+	struct nrf_rpc_ipc *ipc_config = transport->ctx;
+	struct nrf_rpc_ipc_endpoint *endpoint = &ipc_config->endpoint;
+
+	if (!ipc_config->used) {
+		LOG_ERR("nRF RPC transport is not initialized");
+		return -EFAULT;
+	}
+
+	LOG_DBG("Sending %u bytes", length);
+	DUMP_LIMITED_DBG(data, length, "Data: ");
+
+	err = ipc_service_send_nocopy(&endpoint->ept, data, length);
+	if (err < 0) {
+		LOG_ERR("ipc_service_send_nocpy returned err: %d", err);
+		LOG_ERR("Dropping Tx buffer");
+
+		err2 = ipc_service_drop_tx_buffer(&endpoint->ept, data);
```

Comment:
Currently `ipc_rpmsg_static_vrings` backend does not support drop the buffer: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c#L577-L582
Doesn't this cause a potential issue?

Reply:
It might be. It looks like there is outgoing work on adding support for it in the OpenAMP: https://github.com/OpenAMP/open-amp/issues/361